binary_search_tree/tree.py

```diff
@@ -14,44 +14,166 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity:  O(log n)
+    # Space Complexity: O(log n)
+
+    def add_helper(self, current_node, key, value):
+        if current_node == None:
+            return TreeNode(key, value)
+        
+        if key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
+
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+    
+        # Non-recursive version
+        # if self.root == None:
+        #     self.root = TreeNode(key, value)
+        # else:
+        #     parent = None
+        #     current = self.root
+        #     while current != None:
+        #         parent = current
+        #         if current.key > key:
+        #             current = current.left
+        #         else:
+        #             current = current.right
 
-    # Time Complexity: 
-    # Space Complexity: 
+        #     if parent.key > key:
+        #         parent.left = TreeNode(key, value)
+        #     else:
+        #         parent.right = TreeNode(key, value)
+        
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+        
+        return None
+
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(n)
+
+    def inorder_helper(self, root, node_list):
```

Comment:
👍 , Time and space are both O(n) because you're visiting each node and building a list of all the nodes.